lib/Sema/TypeCheckStmt.cpp

```diff
@@ -800,31 +818,41 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
                                        diag::iterator_protocol_broken);
     if (!elementTy)
       return nullptr;
-    
-    // Compute the expression that advances the generator.
-    Expr *iteratorNext
-      = TC.callWitness(TC.buildCheckedRefExpr(generator, DC,
-                                              DeclNameLoc(S->getInLoc()),
-                                              /*implicit*/true),
-                       DC, generatorProto, *genConformance,
-                       TC.Context.Id_next, {}, diag::iterator_protocol_broken);
-    if (!iteratorNext) return nullptr;
-    // Check that next() produces an Optional<T> value.
-    if (iteratorNext->getType()->getAnyNominal()
-          != TC.Context.getOptionalDecl()) {
-      TC.diagnose(S->getForLoc(), diag::iterator_protocol_broken);
+
+    auto *varRef =
+        TC.buildCheckedRefExpr(generator, DC, DeclNameLoc(S->getInLoc()),
+                               /*implicit*/ true);
+    if (!varRef)
       return nullptr;
-    }
+
+    S->setGeneratorVarRef(varRef);
+
+    auto witness =
+        genConformance->getWitnessByName(generatorTy, TC.Context.Id_next);
+    S->setGeneratorNext(witness);
+
+    auto nextResultType =
```

Comment:
Sorry to nitpick this line again -- but now I think you can use `getTypeWitnessByName()`, wrapping the result in an OptionalType, and it will be simpler